Common/src/main/groovy/io/github/groovymc/cgl/api/datagen/recipe/GShapedRecipeBuilder.groovy

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2022 GroovyMC and contributors
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+
+package io.github.groovymc.cgl.api.datagen.recipe
+
+import com.google.common.collect.Lists
+import com.google.common.collect.Maps
+import com.google.common.collect.Sets
+import com.google.gson.JsonArray
+import com.google.gson.JsonObject
+import groovy.transform.CompileStatic
+import net.minecraft.advancements.Advancement
+import net.minecraft.advancements.AdvancementRewards
+import net.minecraft.advancements.RequirementsStrategy
+import net.minecraft.advancements.critereon.RecipeUnlockedTrigger
+import net.minecraft.core.registries.BuiltInRegistries
+import net.minecraft.data.recipes.CraftingRecipeBuilder
+import net.minecraft.data.recipes.FinishedRecipe
+import net.minecraft.resources.ResourceLocation
+import net.minecraft.tags.TagKey
+import net.minecraft.world.item.Item
+import net.minecraft.world.item.ItemStack
+import net.minecraft.world.item.crafting.CraftingBookCategory
+import net.minecraft.world.item.crafting.Ingredient
+import net.minecraft.world.item.crafting.RecipeSerializer
+import net.minecraft.world.level.ItemLike
+import org.jetbrains.annotations.Nullable
+
+import java.util.function.Consumer
+
+@CompileStatic
+class GShapedRecipeBuilder extends CraftingRecipeBuilder implements SimpleRecipeBuilder<GShapedRecipeBuilder> {
+    private final List<String> rows = Lists.newArrayList()
+    private final Map<Character, Ingredient> key = Maps.newLinkedHashMap()
+    private boolean showNotification = true
+
+    GShapedRecipeBuilder define(String symbol, TagKey<Item> tag) {
+        this.define(symbol, Ingredient.of(tag))
+    }
+
+    GShapedRecipeBuilder define(String symbol, ItemLike item) {
+        this.define(symbol, Ingredient.of(item))
+    }
+
+    GShapedRecipeBuilder define(String symbol, Ingredient ingredient) {
+        if (this.key.containsKey(symbol)) {
+            throw new IllegalArgumentException("Symbol '$symbol' is already defined!")
+        } else if (symbol == ' ') {
+            throw new IllegalArgumentException("Symbol ' ' (whitespace) is reserved and cannot be defined")
+        } else {
+            this.key.put(symbol.charAt(0), ingredient)
+            return this
+        }
+    }
+
+    GShapedRecipeBuilder pattern(String... patterns) {
+        patterns.each { pattern ->
```

Comment:
Detecting a single multiline string and transforming that into separate strings by trimming and splitting would be great for Groovy ergonomics.
```groovy
assert pattern '''
   x..
   .x.
   ..x
''' == pattern 'x..', '.x.', '..x'
```

Reply:
That is not possible, considering that spaces which can be trimmed are air ingredients.

Reply:
It is possible. You would trimIndent() first then split by line. This shouldn't interfere with air ingredients unless you have a line of air on the left column, in which case the user defining a character such as A for air similar to other ingredients would work.

Reply:
Yeah, trimming the indent seems fine to me, honestly.